c++/src/kj/async.c++

```diff
@@ -914,16 +917,31 @@ void FiberBase::switchToMain() {
 }
 
 void FiberBase::run() {
+  bool caughtCanceled = false;
   state = RUNNING;
   KJ_DEFER(state = FINISHED);
 
   WaitScope waitScope(currentEventLoop(), *this);
+  
   KJ_IF_MAYBE(exception, kj::runCatchingExceptions([&]() {
-    runImpl(waitScope);
+    try {
+      runImpl(waitScope);
+    } catch (CanceledException) {
+      if (state != CANCELED) {
+        // no idea who would throw this but it's not really our problem
+        throw;
+      }
+      caughtCanceled = true;
+    }
   })) {
     result.addException(kj::mv(*exception));
   }
 
+  if (state == CANCELED && !caughtCanceled) {
+    // if the fiber caught the canceled exception, throw it again
```

Comment:
Throwing here will terminate the process, which is a little too harsh. I was thinking something like:

    KJ_LOG(ERROR, "Canceled fiber apparently caught CanceledException and didn't rethrow it. "
        "Generally, applications should not catch CanceledException, but if they do, they must always rethrow.");